### PR TITLE
evalengine: implement AggregateEvalTypes

### PR DESCRIPTION
**Pull request by vmg**
## Description

As requested by @systay in https://github.com/vitessio/vitess/pull/15069#pullrequestreview-1850341374, this PR implements a helper in the `evalengine` that performs type aggregation on the engine's native `Type` struct, as opposed to `sqltypes.Type` which is what we were previously using.

cc @dbussink 

## Related Issue(s)

- https://github.com/vitessio/vitess/pull/15069#pullrequestreview-1850341374
- 
## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [x] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation was added or is not required

### CODE REVIEW

**Review by dbussink — COMMENTED** [review]
I think we should have some tests for this as well?

**Comment by vmg** [comment]
Tests added, and whattdoyaknow they even caught a missing TODO that I've just implemented. :)

File: go/vt/vtgate/evalengine/api_type_aggregation_test.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
+package evalengine
```
Comment (systay) [inline go/vt/vtgate/evalengine/api_type_aggregation_test.go:1]: 🕺 License yo

**Review by systay — APPROVED** [review]

**Review by GuptaManan100 — APPROVED** [review]

File: go/vt/vtgate/evalengine/api_type_aggregation.go [hunk 2]
```diff
@@ -128,6 +153,23 @@ func (ta *typeAggregation) add(tt sqltypes.Type, f typeFlag) {
...
+	case sqltypes.Uint8:
+		return sqltypes.Int16
+	case sqltypes.Uint16:
+		return sqltypes.Int24
+	case sqltypes.Uint24:
+		return sqltypes.Int32
+	case sqltypes.Uint32:
+		return sqltypes.Int64
+	case sqltypes.Uint64:
+		return sqltypes.Decimal
```
Comment (dbussink) [inline go/vt/vtgate/evalengine/api_type_aggregation.go:167]: Anything where this happens in practice? Wouldn’t instead MySQL fail with an overflow error if this happens?

**Review by dbussink — APPROVED** [review]
One question I had but doesn’t have to be blocking.

File: go/vt/vtgate/evalengine/api_type_aggregation.go (hunk 2 above)
Comment (vmg) [reply go/vt/vtgate/evalengine/api_type_aggregation.go:167]: No, this is not an overflow, this is explicitly listed in the refman:

> If there is a combination of signed and unsigned integer types, the result is signed and the precision may be higher. For example, if the types are signed [INT](https://dev.mysql.com/doc/refman/8.0/en/integer-types.html) and unsigned [INT](https://dev.mysql.com/doc/refman/8.0/en/integer-types.html), the result is signed [BIGINT](https://dev.mysql.com/doc/refman/8.0/en/integer-types.html).

> The exception is unsigned [BIGINT](https://dev.mysql.com/doc/refman/8.0/en/integer-types.html) combined with any signed integer type. The result is [DECIMAL](https://dev.mysql.com/doc/refman/8.0/en/fixed-point-types.html) with sufficient precision and scale 0.

**Pull request merged**